src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+        const length = context.endPosition === undefined ? 0 : context.endPosition - context.startPosition;
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length });
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+        const extractions = extractRange(targetRange, context.file, context);
+
+        let i = 0;
+        for (const extr of extractions) {
+            const name = `scope_${i}`;
+            if (name === actionName) {
+                return ({
+                    edits: extr.changes
+                });
+            }
+            i++;
+        }
+        // ?
+        return undefined;
+    }
+
+    // Move these into diagnostic messages if they become user-facing
+    namespace Messages {
+        function m(message: string): DiagnosticMessage {
+            return { message, code: 0, category: DiagnosticCategory.Message, key: message };
+        }
+
+        export const CannotExtractFunction: DiagnosticMessage = m("Cannot extract function.");
+        export const StatementOrExpressionExpected: DiagnosticMessage = m("Statement or expression expected.");
+        export const CannotExtractRangeContainingConditionalBreakOrContinueStatements: DiagnosticMessage = m("Cannot extract range containing conditional break or continue statements.");
+        export const CannotExtractRangeContainingConditionalReturnStatement: DiagnosticMessage = m("Cannot extract range containing conditional return statement.");
+        export const CannotExtractRangeContainingLabeledBreakOrContinueStatementWithTargetOutsideOfTheRange: DiagnosticMessage = m("Cannot extract range containing labeled break or continue with target outside of the range.");
+        export const CannotExtractRangeThatContainsWritesToReferencesLocatedOutsideOfTheTargetRangeInGenerators: DiagnosticMessage = m("Cannot extract range containing writes to references located outside of the target range in generators.");
+        export const TypeWillNotBeVisibleInTheNewScope = m("Type will not visible in the new scope.");
+        export const FunctionWillNotBeVisibleInTheNewScope = m("Function will not visible in the new scope.");
+        export const InsufficientSelection = m("Select more than a single identifier.");
+    }
+
+    export enum RangeFacts {
+        None = 0,
+        HasReturn = 1 << 0,
+        IsGenerator = 1 << 1,
+        IsAsyncFunction = 1 << 2,
+        UsesThis = 1 << 3
+    }
+
+    /**
+     * Represents an expression or a list of statements that should be extracted with some extra information
+     */
+    export interface TargetRange {
+        readonly range: Expression | Statement[];
+        readonly facts: RangeFacts;
+    }
+
+    /**
+     * Result of 'getRangeToExtract' operation: contains either a range or a list of errors
+     */
+    export interface RangeToExtract {

```

Comment:
With structural typing, I think of a name as shorthand for a description of the type.  Introducing shorthand to shorten exactly one instance seems counterproductive.  I don't feel strongly though - names can also be explanatory.